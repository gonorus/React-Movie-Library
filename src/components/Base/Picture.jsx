import React, { useState } from 'react';
import PropTypes from 'prop-types';
import './Picture.scss';

import Spinner from '../Misc/Spinner';

/**
 * Picture Component with Lazy image loading behavior
 */
const Picture = React.forwardRef(({ src, alt }, ref) => {
  const [isLoaded, setIsLoaded] = useState(false);
  return (
    <div className='picture-container'>
      <img
        ref={ref}
        loading='lazy'
        src={src}
        alt={alt}
        onLoad={() => setIsLoaded(true)}
        onError={() => setIsLoaded(true)}
      />
      { isLoaded === false && <Spinner /> }
    </div>
  );
});
Picture.propTypes = {
  src: PropTypes.string.isRequired,
  alt: PropTypes.string,
};
Picture.defaultProps = {
  alt: undefined,
};
export default Picture;
