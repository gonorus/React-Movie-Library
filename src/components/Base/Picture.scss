.picture-container {
  width: 100%;
  height: 100%;
  display: grid;
  place-content: center;

  img {
    width: 100%;
    height: 100%;
  }

  .showbox {
    position: unset;
  }
}