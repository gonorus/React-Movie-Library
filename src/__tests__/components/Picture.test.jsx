import React from 'react';
import renderer, { act } from 'react-test-renderer';
import Picture from '../../components/Base/Picture';

test('Picture loaded', () => {
  const component = renderer.create(
    <Picture src='https://images.unsplash.com/photo-1637249805971-59d7b9319df3?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1740&q=80' alt='alternate image' />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  const componentInstance = component.root;
  const imgInstance = componentInstance.findByType('img');

  act(() => imgInstance.props.onLoad());
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  act(() => imgInstance.props.onError());
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
