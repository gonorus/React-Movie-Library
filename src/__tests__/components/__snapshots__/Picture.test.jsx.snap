// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Picture loaded 1`] = `
<div
  className="picture-container"
>
  <img
    alt="alternate image"
    loading="lazy"
    onError={[Function]}
    onLoad={[Function]}
    src="https://images.unsplash.com/photo-1637249805971-59d7b9319df3?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1740&q=80"
  />
  <div
    className="showbox"
  >
    <div
      className="loader"
      style={
        Object {
          "width": 50,
        }
      }
    >
      <svg
        className="circular"
        viewBox="25 25 50 50"
      >
        <circle
          className="path"
          cx="50"
          cy="50"
          fill="none"
          r="20"
          strokeMiterlimit="10"
          strokeWidth="2"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Picture loaded 2`] = `
<div
  className="picture-container"
>
  <img
    alt="alternate image"
    loading="lazy"
    onError={[Function]}
    onLoad={[Function]}
    src="https://images.unsplash.com/photo-1637249805971-59d7b9319df3?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1740&q=80"
  />
</div>
`;

exports[`Picture loaded 3`] = `
<div
  className="picture-container"
>
  <img
    alt="alternate image"
    loading="lazy"
    onError={[Function]}
    onLoad={[Function]}
    src="https://images.unsplash.com/photo-1637249805971-59d7b9319df3?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1740&q=80"
  />
</div>
`;
